Splice: Interop Final (probably) with Eclair

ddustin opened the pull request:
Interop testing with Eclair revealed an issue with remote funding key rotation.

This searches for the funding output using the rotated remote funding pubkey instead of the current funding pubkey.

Update the variable name to be more clear which this represents.

Implement new pending spec changes for `splice_locked` resuming in reestablish.

Fixes #8030

ddustin commented:
@remyers reported this fix made his splice interop test pass! 🎉🚀

niftynei commented on channeld/channeld.c:
```diff
@@ -3829,7 +3829,7 @@ static void splice_accepter(struct peer *peer, const u8 *inmsg)
 	new_inflight->outpoint = outpoint;
 	new_inflight->amnt = both_amount;
-	new_inflight->psbt = tal_steal(new_inflight, ictx->current_psbt);
+	new_inflight->psbt = clone_psbt(new_inflight, ictx->current_psbt);
```
yeah tal_steal and psbt code aren't good friends

ddustin replied:
No more theft among psbt 🙅

niftynei commented on channeld/channeld.c:
```diff
@@ -511,13 +511,14 @@ static void check_mutual_splice_locked(struct peer *peer)
+	struct bitcoin_txid splice_txid;
 
-	if (!fromwire_splice_locked(msg, &chanid))
+	if (!fromwire_splice_locked(msg, &chanid, &splice_txid))
```
we're not checking it?

ddustin replied:
Added a check for it in [e1d7602](https://github.com/ElementsProject/lightning/pull/8021/commits/e1d760245d689163efd73b61f93e2d5d99dd4b39)

niftynei reviewed (commented):
very cool!

ddustin commented on lightningd/channel.c:
```diff
@@ -200,6 +200,7 @@ new_inflight(struct channel *channel,
 	inflight->i_am_initiator = i_am_initiator;
 	inflight->force_sign_first = force_sign_first;
+	inflight->is_locked = false;
 	inflight->splice_locked_memonly = false;
```
Added

niftynei reviewed (approved):
ACK e91b6b796e80c34e793d103718a11df7a1a25106

pending tests pass :)

endothermicdev commented:
Removed extra whitespace to check how CI is progressing otherwise.

rustyrussell commented:
Rebased on master which will fix some broken tests...

remyers commented:
The interop test [successful-single-splice](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#successful-single-splice) is failing:
- eclair opens a channel with cln
- eclair initiates a splice-in of 50000 sats, but splice does not lock
- eclair pays invoice from cln for 10000 sats
- splice locks in (never reached)

This results in the following error and a disconnect:
```
 billboard perm: Bad commit_sig signature 1 <cut> for tx  <cut> wscript <cut> feerate 316. Cur funding <cut>, splice_info: <cut>, race_await_commit: no, inflight splice count: 1
```
[single_splice_cln.log](https://github.com/user-attachments/files/19336743/single_splice_cln.log)

remyers commented:
In my testing of [Multiple splices with concurrent splice_locked](https://github.com/lightning/bolts/blob/4e82adcd848466b36a6d8a717b195eabe24de216/bolt02/splicing-test.md#multiple-splices-with-concurrent-splice_locked) from the splice test vectors I found that the `tx_init_rbf` message is rejected.
```log
2025-04-09T15:42:31.982Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: peer_in WIRE_STFU
2025-04-09T15:42:31.982Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: STFU initiator was remote.
2025-04-09T15:42:31.982Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: Sending peer that we want to STFU.
2025-04-09T15:42:31.982Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: peer_out WIRE_STFU
2025-04-09T15:42:31.982Z UNUSUAL 02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: STFU complete: we are quiescent
2025-04-09T15:42:31.982Z UNUSUAL 02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: STFU complete: setting stfu_wait_single_msg = true
2025-04-09T15:42:31.988Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: peer_in WIRE_TX_INIT_RBF
2025-04-09T15:42:31.988Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: peer_out WIRE_WARNING
2025-04-09T15:42:31.989Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: billboard perm: Got invalid message during STFU mode: WIRE_TX_INIT_RBF
2025-04-09T15:42:31.989Z INFO    02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-chan#1: Peer transient failure in CHANNELD_AWAITING_SPLICE: channeld sent Got invalid message during STFU mode: WIRE_TX_INIT_RBF
2025-04-09T15:42:31.989Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-channeld-chan#1: Status closed, but not exited. Killing
2025-04-09T15:42:31.989Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-connectd: disconnect_peer
2025-04-09T15:42:31.989Z DEBUG   connectd: drain_peer
2025-04-09T15:42:31.989Z DEBUG   connectd: drain_peer draining subd!
2025-04-09T15:42:31.989Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-lightningd: peer_disconnect_done
2025-04-09T15:42:31.989Z DEBUG   02b9b6992d9c1eb1681939d76eea77565b1b61f57f7a9c49d12598eb2d95a5366a-connectd: Will try reconnect in 1 seconds
```

remyers commented:
I've found an issue with the interop test: "[Disconnection with both sides sending commit_sig](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-commit_sig)"

When Eclair and CLN send their `channel_reestablish` messages, both nodes correctly set `next_funding_txid` to the same txid and  Eclair correctly sets `next_commit_number` to 0, which triggers CLN to correctly resend their `commit_sig` message. But CLN sets their `next_commit_number` to 1, and so Eclair never resends their `commit_sig` to CLN because Eclair's last sent `commit_sig` had commitment number 0 and so does not match. 

The relevant part of the splice spec is [here](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L3392):
```md
A receiving node:
  - if `next_funding_txid` is set:
    - if `next_funding_txid` matches the latest interactive funding transaction
      or the current channel funding transaction:
      - if `next_commitment_number` is equal to the commitment number of the
        `commitment_signed` message it sent for this funding transaction:
        - MUST retransmit its `commitment_signed` for that funding transaction.

```
[cln.log](https://github.com/user-attachments/files/19910860/cln.log)
[eclair.log](https://github.com/user-attachments/files/19910874/eclair.log) [see lines 189 and 199]

remyers commented:
I've found an issue with the interop test: "[Disconnection with one side sending tx_signatures](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-one-side-sending-tx_signatures)"

When reconnecting CLN sends a `channel_reestablish` messages with the `my_current_funding_locked` TLV that is the same as `next_funding_txid`, even though they have not received `tx_signatures` yet from Eclair (nor has the tx been published). The correct behavior would be for `my_current_funding_locked` to be the same as `your_last_funding_locked`, which is the original funding txid - not the splice txid.

The relevant part of the splice spec is [here](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L3324):
```
   - if a splice transaction reached acceptable depth while disconnected:
     ...
   - otherwise:
      - MUST set `my_current_funding_locked` to the txid of the last `splice_locked` it sent.
```
[cln.log](https://github.com/user-attachments/files/19966414/cln.log)
```
their channel_reestablish: next_funding_tx_id=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), your_last_funding_locked=Some(3962b0373e4775f9c082f63963b138b67475ddf8036f34086ad9cd3f3d8725a2), my_current_funding_Locked=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), next_local_commit_number=1
```

remyers commented:
I've found an issue with the interop test: "[Disconnection with one side sending tx_signatures](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-one-side-sending-tx_signatures)"

When reconnecting CLN sends a `channel_reestablish` messages with the `my_current_funding_locked` TLV that is the same as `next_funding_txid`, even though they have not received `tx_signatures` yet from Eclair (nor has the tx been published). The correct behavior would be for `my_current_funding_locked` to be the same as `your_last_funding_locked`, which is the original funding txid - not the splice txid.

The relevant part of the splice spec is [here](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L3324):
```
   - if a splice transaction reached acceptable depth while disconnected:
     ...
   - otherwise:
      - MUST set `my_current_funding_locked` to the txid of the last `splice_locked` it sent.
```
[cln.log](https://github.com/user-attachments/files/19966414/cln.log)
```
their channel_reestablish: next_funding_tx_id=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), your_last_funding_locked=Some(3962b0373e4775f9c082f63963b138b67475ddf8036f34086ad9cd3f3d8725a2), my_current_funding_Locked=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), next_local_commit_number=1
```

remyers commented:
> I've found an issue when re-running the interop test: "[Disconnection with both sides sending commit_sig](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-commit_sig)" after rebasing on your latest changes in [facfd7b23](https://github.com/ddustin/lightning/commit/facfd7b234b5a7a05eff7e9b911fc8ec5c5ccbc4).
```
> ...
>     //   |<---- tx_signatures ---|
>     //   |----- tx_signatures -->|
>     //   |<---- splice_locked ---|
>     //   |<----- commit_sig -----|  *** signing original funding tx, invalid signature, see line 538 of cln.log
>     //   |<----- commit_sig -----| *** signing splice funding tx
> ```


I'll investigate why Eclair is not ignoring those redundant commit_sigs, especially if one of them is for the previous funding tx which has been replaced by the new splice tx.

ddustin commented:
> I've found an issue with the interop test: "[Disconnection with one side sending tx_signatures](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-one-side-sending-tx_signatures)"
> 
> When reconnecting CLN sends a `channel_reestablish` messages with the `my_current_funding_locked` TLV that is the same as `next_funding_txid`, even though they have not received `tx_signatures` yet from Eclair (nor has the tx been published). The correct behavior would be for `my_current_funding_locked` to be the same as `your_last_funding_locked`, which is the original funding txid - not the splice txid.
> 
> The relevant part of the splice spec is [here](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L3324):
> 
> ```
>    - if a splice transaction reached acceptable depth while disconnected:
>      ...
>    - otherwise:
>       - MUST set `my_current_funding_locked` to the txid of the last `splice_locked` it sent.
> ```
> 
> [cln.log](https://github.com/user-attachments/files/19966414/cln.log) [eclair.log](https://github.com/user-attachments/files/19966465/eclair.log) [see line 215]
> 
> ```
> their channel_reestablish: next_funding_tx_id=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), your_last_funding_locked=Some(3962b0373e4775f9c082f63963b138b67475ddf8036f34086ad9cd3f3d8725a2), my_current_funding_Locked=Some(d7c41bf0ea79ab0c0bba3bc9ee4809cd6242bda6d5f0d0035bbbd8afe163c255), next_local_commit_number=1
> ```

I can't see how this is possible so I added more log messages to add clarity of what CLN is sending. Can you re-run this test and get the new logs? f1549d436378ff7c83dc80aee73254a8c2db2bf4

ddustin commented:
> I've found an issue when re-running the interop test: "[Disconnection with both sides sending commit_sig](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-commit_sig)" after rebasing on your latest changes in [facfd7b23](https://github.com/ddustin/lightning/commit/facfd7b234b5a7a05eff7e9b911fc8ec5c5ccbc4).
> 
> The first problem was that it failed at this check in [channeld.c](https://github.com/ElementsProject/lightning/blob/facfd7b234b5a7a05eff7e9b911fc8ec5c5ccbc4/channeld/channeld.c#L5596) because `next_commitment_number` is 0:
> 
> ```c
> 	if (next_commitment_number == peer->next_index[REMOTE] - 1) {
>                /* We completed opening, we don't re-transmit that one! */
>                if (next_commitment_number == 0)
> 			peer_failed_err(peer->pps,
> 					 &peer->channel_id,
> 					 "bad reestablish commitment_number is 0 when no splice inflight");
> 		retransmit_commitment_signed = true;
> ```
> 
> I modified the check to instead be:
> 
> ```c
> 		/* We completed opening, we don't re-transmit that one unless negotiating a splice. */
> 		if (remote_next_funding == NULL && next_commitment_number == 0)
> ```
> 
> But now I see the following message flow:
> 
> ```
>     // eclair                   cln
>     //   |         ...           |
>     //   |    <interactive-tx>   |
>     //   |<----- tx_complete ----|
>     //   |------ tx_complete --->|
>     //   |------ commit_sig ---X |
>     //   |<------ commit_sig ----|
>     //   |      <disconnect>     |
>     //   |      <reconnect>      |
>     //   | <channel_reestablish> |
>     //   |<----- commit_sig -----|  
>     //   |------ commit_sig ---->|
>     //   |<---- tx_signatures ---|
>     //   |----- tx_signatures -->|
>     //   |<---- splice_locked ---|
>     //   |<----- commit_sig -----|  *** signing original funding tx, invalid signature, see line 538 of cln.log
>     //   |<----- commit_sig -----| *** signing splice funding tx
> ```
> 
> Seems like the [facfd7b23](https://github.com/ddustin/lightning/commit/facfd7b234b5a7a05eff7e9b911fc8ec5c5ccbc4) commit broke this test. [cln.log](https://github.com/user-attachments/files/19976361/cln.log) [eclair.log](https://github.com/user-attachments/files/19976364/eclair.log)

Marking this as completed on my side because of our conversation that CLN is allowed but not required to send the last two `commit_sig` messages so Eclair should be updated to handle them.

ddustin commented:
> In my testing of [Multiple splices with concurrent splice_locked](https://github.com/lightning/bolts/blob/4e82adcd848466b36a6d8a717b195eabe24de216/bolt02/splicing-test.md#multiple-splices-with-concurrent-splice_locked) from the splice test vectors I found that the `tx_init_rbf` message is rejected.

RBF is enabled now (but the tests aren't written yet).

remyers commented:
> Marking this as completed on my side because of our conversation that CLN is allowed but not required to send the last two `commit_sig` messages so Eclair should be updated to handle them.

Could you double check to make sure that CLN is sending the same signatures it sent before? I modified Eclair to `commitment_signature` messages with a valid signature that it has already received. For my local tests this fixes it, but it is till failing when I test against CLN. 

I also noticed after Eclair sends an Error, the CLN log includes "insufficient fee, rejecting replacement" when trying to submit the confirm txs to bitcoind. Could you take a look at that? [cln.log](https://github.com/user-attachments/files/19976361/cln.log) [eclair.log](https://github.com/user-attachments/files/19976364/eclair.log)

I'm going to rebase on your latest changes and see if that changes anything.

remyers commented:
> Edit: Found a way it would be possible and patched it with the latest push.

These tests now work as of [d9da56d](https://github.com/ElementsProject/lightning/commit/d9da56dd72cac584d2f350894a65befa89aae69a):
- [Successful single splice](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#successful-single-splice)
- [Disconnection with one side sending commit_sig](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-one-side-sending-commit_sig)
- [Disconnection with one side sending tx_signatures](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-one-side-sending-tx_signatures)
- [Disconnection with both sides sending tx_signatures](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-tx_signatures)

I still need to test:
- [Multiple splices with concurrent splice_locked](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#multiple-splices-with-concurrent-splice_locked)
- [Disconnection with both sides sending tx_signatures and channel updates](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-tx_signatures-and-channel-updates)
- [Disconnection with concurrent splice_locked](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-concurrent-splice_locked)

Test fails (due to extra commit_sigs):
- [Disconnection with both sides sending commit_sig](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#disconnection-with-both-sides-sending-commit_sig)

remyers commented:
If Eclair sends `tx_abort` after CLN sends `tx_ack_rbf` then CLN will force close the channel with the error message: "tx_abort is not allowed after I have sent my signature."

I found this testing a variant of the interop test: [Multiple splices with concurrent splice_locked](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#multiple-splices-with-concurrent-splice_locked).

The [spec](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L538) allows `tx_abort` as a response to `tx_ack_rbf` and should not cause a force close. I also suggest that any invalid message after `tx_ack_rbf` should result in a disconnect and `channel_reestablish`, not a force close. 

[eclair.log](https://github.com/user-attachments/files/20037446/eclair.log)
[cln.log](https://github.com/user-attachments/files/20037452/cln.log) [see line 1184]

ddustin commented:
> If Eclair sends `tx_abort` after CLN sends `tx_ack_rbf` then CLN will force close the channel with the error message: "tx_abort is not allowed after I have sent my signature."
> 
> I found this testing a variant of the interop test: [Multiple splices with concurrent splice_locked](https://github.com/lightning/bolts/blob/8c907f6b8d26fad8ec79ad1fe3078eb92e5285a6/bolt02/splicing-test.md#multiple-splices-with-concurrent-splice_locked).
> 
> The [spec](https://github.com/lightning/bolts/blob/02da8e45503f41716c133d5f19c6ce8245b1099d/02-peer-protocol.md?plain=1#L538) allows `tx_abort` as a response to `tx_ack_rbf` and should not cause a force close. I also suggest that any invalid message after `tx_ack_rbf` should result in a disconnect and `channel_reestablish`, not a force close.
> 
> [eclair.log](https://github.com/user-attachments/files/20037446/eclair.log) [cln.log](https://github.com/user-attachments/files/20037452/cln.log) [see line 1184]

This was due to an overly aggressive `tx_abort` check. Updated the check to verify the `txid` being aborted specifically and relaxed the case where no inflight has being deleted. This should hopefully fix the issue 👍.

Also addressed the `splice_locked` failure you sent me on DM. That appeared to be a check aimed to prevent multiple redundant txid watches -- removed the check and it should hopefully fix that issue as well.

I think 👀 I have all the interop issues addressed so far now. Let me know if I'm missing any.

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/commit/9b74028c0e23b5105e51f5dae3d45da557670d4c

I've kicked the two (known) flakes...
